src/D_functions_and_procedures.sql

```diff
@@ -21,4 +21,89 @@ $$
 LANGUAGE 'plpgsql';
 
 --D2 TEST
-SELECT * FROM simple_search('apple',1);
\ No newline at end of file
+SELECT * FROM simple_search('apple',1);
+
+
+--D3 Rating function
+CREATE OR REPLACE FUNCTION rate(p_userid INT, p_imdb_id VARCHAR, p_score NUMERIC)
+RETURNS VOID AS $$
+DECLARE
+    v_media_id INTEGER;
+    v_existing_user_score RECORD;
+BEGIN
+    -- Get media ID from IMDb ID
+    SELECT media_id INTO v_media_id FROM media WHERE imdb_id = p_imdb_id;
+
+    -- Check if the user has already rated this media in user_score
+    SELECT * INTO v_existing_user_score 
+    FROM user_score 
+    WHERE user_id = p_userid AND media_id = v_media_id;
+
+    IF FOUND THEN
+        -- Update user's existing score
+        UPDATE user_score
+        SET score_value = p_score,
+            created_at = CURRENT_TIMESTAMP
+        WHERE user_id = p_userid AND media_id = v_media_id;
+    ELSE
+        -- Insert new score into user_score
+        INSERT INTO user_score (user_id, media_id, score_value)
+        VALUES (p_userid, v_media_id, p_score);
+    END IF;
+
+    -- Update the score table with the new average rating
+    -- Check if the movie already has a score entry
+    IF EXISTS (SELECT 1 FROM score WHERE media_id = v_media_id) THEN
+        -- Update the existing average score and vote count in the score table
+        UPDATE score
+        SET value = (SELECT AVG(score_value) FROM user_score WHERE media_id = v_media_id),
+            vote_count = (SELECT COUNT(*) FROM user_score WHERE media_id = v_media_id),
+            "at" = CURRENT_TIMESTAMP
+        WHERE media_id = v_media_id;
```

Comment:
Think you need to also check if source matches `userrating` or you'll end up updating all other scores for that media as well